[ Inductor ] Shape padding honors output stride preservation

kadeng opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119578
* __->__ #120797

This fix makes sure that shape padding honors inductors 'keep_output_strides' setting.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

kadeng committed:
[ Inductor ] Shape padding honors output stride preservation

This fix makes sure that shape padding honors inductors 'keep_output_strides' setting.

[ghstack-poisoned]

kadeng committed:
Update on "[ Inductor ] Shape padding honors output stride preservation"


This fix makes sure that shape padding honors inductors 'keep_output_strides' setting.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

kadeng committed:
Update on "[ Inductor ] Shape padding honors output stride preservation"


This fix makes sure that shape padding honors inductors 'keep_output_strides' setting.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

eellison commented on torch/_inductor/fx_passes/pad_mm.py:
```diff
@@ -108,6 +213,11 @@ def addmm_pattern(
 
 
 def should_pad_addmm(match: Match) -> bool:
+    if (
+        torch._inductor.config.keep_output_stride
+        and _result_layout_affects_graph_output(match)
+    ):
+        return False
```
Maybe I'm missing something but can we just check if the output of the addmm is a graph output ?

kadeng replied:
If we would be on the IR level, I think that would be sufficient if we unpack views. But since we're looking at the FX graph, I imagine the GEMM memory output layout could propagate through node fusions. But I'm not sure if this is an entirely realistic concern, since a GEMM followed by a slice and then followed by a Pointwise op would not be fused with the Pointwise because the slicing means that the Pointwise op should not be applied to the entirety of the GEMM output. So, yes, maybe we can simplify this a lot, but I'm unsure if it's safe at the moment, and pretty sure it's hard to guarantee it stays safe.

eellison replied:
Non-dense format will only propagate to output through views. Check examples here of looking at uses/views https://github.com/pytorch/pytorch/blob/ea7149aa2233b7987aadc3d6e2ba94cd3e133606/torch/_inductor/utils.py#L306-L326.

kadeng replied:
Thanks for the code pointer, that allowed me to simplify it a lot..

kadeng committed:
Update on "[ Inductor ] Shape padding honors output stride preservation"


This fix makes sure that shape padding honors inductors 'keep_output_strides' setting.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

kadeng committed:
Update on "[ Inductor ] Shape padding honors output stride preservation"


This fix makes sure that shape padding honors inductors 'keep_output_strides' setting.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

kadeng committed:
Update on "[ Inductor ] Shape padding honors output stride preservation"


This fix makes sure that shape padding honors inductors 'keep_output_strides' setting.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

kadeng committed:
Update on "[ Inductor ] Shape padding honors output stride preservation"


This fix makes sure that shape padding honors inductors 'keep_output_strides' setting.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

eellison commented on torch/_inductor/fx_passes/pad_mm.py:
```diff
@@ -52,6 +59,50 @@ def check_dtype(a: Tensor, b: Tensor) -> bool:
+    assert search_node is not None
+    assert graph is not None
+
+    def find_output(node: torch.fx.Node, is_start_node=False):
```
nit: add `seen` set so we don't revisit nodes multiple times.

eellison reviewed (approved):
Looks good. As a follow up we could limit to only the user_visible_outputs - (I.e., not saved tensors for the backward) https://github.com/pytorch/pytorch/blob/3239f86a3df133b5977d988324639e0de7af8749/torch/_inductor/graph.py#L207

kadeng committed:
Update on "[ Inductor ] Shape padding honors output stride preservation"


This fix makes sure that shape padding honors inductors 'keep_output_strides' setting.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

kadeng committed:
Update on "[ Inductor ] Shape padding honors output stride preservation"


This fix makes sure that shape padding honors inductors 'keep_output_strides' setting.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

kadeng commented:
@pytorchmergebot merge

williamwen42 commented:
@pytorchbot revert -m "perf regression on dashboard" -c nosignal
